File: cavity.py
# Lid-driven cavity 2D — vorticidade–função-corrente (Re=100)
# Implementação 100% NumPy (sem Numba). dt adaptativo por CFL.
# Uso: python cavity.py --N 256 --Re 100 --CFL 0.3 --tol 1e-6 --jiters 80

import argparse, os, time, csv
import numpy as np

ap = argparse.ArgumentParser()
ap.add_argument("--N", type=int, default=256)
ap.add_argument("--Re", type=float, default=100.0)
ap.add_argument("--CFL", type=float, default=0.3)
ap.add_argument("--tol", type=float, default=1e-6)
ap.add_argument("--jiters", type=int, default=80)   # iterações de Jacobi por passo
ap.add_argument("--maxiter", type=int, default=20000)
args = ap.parse_args()

N, Re, CFL, tol, JITERS, maxiter = args.N, args.Re, args.CFL, args.tol, args.jiters, args.maxiter
L = 1.0
dx = L/(N-1); dy = dx
nu = 1.0/Re
Utop = 1.0

omega = np.zeros((N,N), dtype=np.float64)
psi   = np.zeros((N,N), dtype=np.float64)
u     = np.zeros((N,N), dtype=np.float64)
v     = np.zeros((N,N), dtype=np.float64)
omega_new = np.zeros_like(omega)

def set_psi_bcs(psi):
    psi[:,0] = 0.0; psi[:,-1] = 0.0
    psi[0,:] = 0.0; psi[-1,:] = 0.0

def update_uv_from_psi(u, v, psi):
    # interior
    u[1:-1,1:-1] = (psi[1:-1,2:] - psi[1:-1,:-2])/(2*dy)
    v[1:-1,1:-1] = -(psi[2:,1:-1] - psi[:-2,1:-1])/(2*dx)
    # paredes (no-slip + tampa U=1)
    u[:,0] = 0.0; v[:,0] = 0.0
    u[:,-1] = Utop; v[:,-1] = 0.0
    u[0,:] = 0.0; v[0,:] = 0.0
    u[-1,:] = 0.0; v[-1,:] = 0.0

def update_wall_vorticity(omega, psi):
    omega[:,0]  = -2.0*psi[:,1]/(dy*dy)                         # bottom
    omega[:,-1] = -2.0*psi[:,-2]/(dy*dy) - 2.0*Utop/dy          # top
    omega[0,:]  = -2.0*psi[1,:]/(dx*dx)                         # left
    omega[-1,:] = -2.0*psi[-2,:]/(dx*dx)                        # right

def step_vorticity(omega, u, v, dt):
    dom = (slice(1,-1), slice(1,-1))
    om = omega
    adv = ( u[dom]*(om[2:,1:-1] - om[:-2,1:-1])/(2*dx)
          + v[dom]*(om[1:-1,2:] - om[1:-1,:-2])/(2*dy) )
    lap = ( (om[2:,1:-1] - 2*om[1:-1,1:-1] + om[:-2,1:-1])/(dx*dx)
          + (om[1:-1,2:] - 2*om[1:-1,1:-1] + om[1:-1,:-2])/(dy*dy) )
    new = om.copy()
    new[dom] = om[dom] + dt*(-adv + nu*lap)
    return new

def jacobi_psi(psi, omega, iters):
    dx2 = dx*dx; dy2 = dy*dy
    beta = 1.0/(2*(dx2+dy2))
    psi_new = np.empty_like(psi)
    for _ in range(iters):
        set_psi_bcs(psi)
        psi_new[1:-1,1:-1] = beta*( (psi[2:,1:-1] + psi[:-2,1:-1])*dy2
                                   + (psi[1:-1,2:] + psi[1:-1,:-2])*dx2
                                   + dx2*dy2*omega[1:-1,1:-1] )
        psi_new[:,0] = 0.0; psi_new[:,-1] = 0.0
        psi_new[0,:] = 0.0; psi_new[-1,:] = 0.0
        psi, psi_new = psi_new, psi
    return psi

t0 = time.perf_counter()
steps = 0
dt_acc = 0.0
res = np.inf

# warm-up leve
set_psi_bcs(psi); update_uv_from_psi(u,v,psi); update_wall_vorticity(omega,psi)
omega_new = step_vorticity(omega, u, v, 1e-6)
psi = jacobi_psi(psi, omega, 1)

for it in range(1, maxiter+1):
    set_psi_bcs(psi)
    update_uv_from_psi(u, v, psi)
    update_wall_vorticity(omega, psi)

    umax = max(np.max(np.abs(u)), 1e-12)
    vmax = max(np.max(np.abs(v)), 1e-12)
    dtc1 = dx/umax; dtc2 = dy/vmax
    dtd1 = dx*dx/(4*nu); dtd2 = dy*dy/(4*nu)
    dt = CFL * min(dtc1, dtc2, dtd1, dtd2)

    omega_new = step_vorticity(omega, u, v, dt)

    # alinhar bordas antes de medir resíduo
    omega_new[0,:]  = omega[0,:]
    omega_new[-1,:] = omega[-1,:]
    omega_new[:,0]  = omega[:,0]
    omega_new[:,-1] = omega[:,-1]
    res = np.max(np.abs(omega_new - omega))

    omega = omega_new
    psi = jacobi_psi(psi, omega, JITERS)

    steps += 1
    dt_acc += dt

    if it % 200 == 0:
        print(f"Iter {it}, dt={dt:.2e}, res={res:.2e}")
    if res < tol:
        print(f"Convergiu em {it} passos")
        break

elapsed = time.perf_counter() - t0
cells = (N-2)*(N-2)
mlups_vort = cells*steps/elapsed/1e6
avg_dt = dt_acc/max(steps,1)

print(f"N={N} Re={Re:.1f} steps={steps} time={elapsed:.3f}s MLUPS(vort)={mlups_vort:.2f} "
      f"avg_dt={avg_dt:.2e} res={res:.2e}")

# perfis centrais (nomes iguais aos da Julia/Python Numba para o run.py usar)
y = np.linspace(0.0, 1.0, N); x = y
np.savetxt(f"u_center_python_N{N}.dat", np.c_[y, u[N//2,:]])
np.savetxt(f"v_center_python_N{N}.dat", np.c_[x, v[:,N//2]])

# resumo CSV (append) — usa summary_python.csv para uniformizar
sumfile = "summary_python.csv"
newfile = not os.path.exists(sumfile)
with open(sumfile, "a", newline="") as f:
    import csv
    w = csv.writer(f)
    if newfile:
        w.writerow(["timestamp","N","Re","CFL","tol","jacobi_iters","steps","time_s","avg_dt","MLUPS_vort","res"])
    w.writerow([time.strftime("%Y-%m-%dT%H:%M:%S"),
                N, Re, CFL, tol, JITERS, steps, f"{elapsed:.6f}", f"{avg_dt:.6e}",
                f"{mlups_vort:.3f}", f"{res:.3e}"])

File: cavity.jl
# Lid-driven cavity 2D — vorticidade–funcao-corrente (Re=100), versao VETORIZADA
# Uso: julia -O3 --check-bounds=no cavity.jl [N] [Re] [CFL] [tol] [JACOBI_ITERS]
# Ex.:  julia -O3 --check-bounds=no cavity.jl 256 100 0.3 1e-6 80

using Printf, Dates

function main()
    # ---------------- CLI ----------------
    N    = length(ARGS) >= 1 ? parse(Int,     ARGS[1]) : 256
    Re   = length(ARGS) >= 2 ? parse(Float64, ARGS[2]) : 100.0
    CFL  = length(ARGS) >= 3 ? parse(Float64, ARGS[3]) : 0.3
    tol  = length(ARGS) >= 4 ? parse(Float64, ARGS[4]) : 1e-6
    JIT  = length(ARGS) >= 5 ? parse(Int,     ARGS[5]) : 80

    L  = 1.0
    dx = L/(N-1);  dy = dx
    nu = 1.0/Re
    Utop = 1.0
    maxiter = 20_000

    # --------------- campos ---------------
    omega    = zeros(N, N)
    psi      = zeros(N, N)
    u        = zeros(N, N)
    v        = zeros(N, N)
    omeganew = similar(omega)
    psinew   = similar(psi)   # buffer para Jacobi vetorizado

    @views begin
        # ---------- helpers ----------
        set_psi_bcs!(psi) = (psi[:,1] .= 0.0; psi[:,end] .= 0.0; psi[1,:] .= 0.0; psi[end,:] .= 0.0)

        function update_uv_from_psi!(u, v, psi)
            u[2:end-1,2:end-1] .= (psi[2:end-1,3:end] .- psi[2:end-1,1:end-2]) ./ (2dy)
            v[2:end-1,2:end-1] .= -(psi[3:end,2:end-1] .- psi[1:end-2,2:end-1]) ./ (2dx)
            # paredes (no-slip; tampa U=1)
            u[:,1]   .= 0.0;  v[:,1]   .= 0.0
            u[:,end] .= Utop; v[:,end] .= 0.0
            u[1,:]   .= 0.0;  v[1,:]   .= 0.0
            u[end,:] .= 0.0;  v[end,:] .= 0.0
        end

        function update_wall_vorticity!(omega, psi)
            omega[:,1]   .= -2 .* psi[:,2]       ./ (dy^2)               # bottom
            omega[:,end] .= -2 .* psi[:,end-1]   ./ (dy^2) .- 2Utop/dy   # top (U=1)
            omega[1,:]   .= -2 .* psi[2,:]       ./ (dx^2)               # left
            omega[end,:] .= -2 .* psi[end-1,:]   ./ (dx^2)               # right
        end

        # omega-step (vetorizado)
        function step_vorticity!(omeganew, omega, u, v, dt)
            adv = u[2:end-1,2:end-1] .* (omega[3:end,2:end-1] .- omega[1:end-2,2:end-1]) ./ (2dx) .+
                  v[2:end-1,2:end-1] .* (omega[2:end-1,3:end] .- omega[2:end-1,1:end-2]) ./ (2dy)
            lap = (omega[3:end,2:end-1] .- 2 .* omega[2:end-1,2:end-1] .+ omega[1:end-2,2:end-1]) ./ (dx^2) .+
                  (omega[2:end-1,3:end] .- 2 .* omega[2:end-1,2:end-1] .+ omega[2:end-1,1:end-2]) ./ (dy^2)
            omeganew .= omega
            omeganew[2:end-1,2:end-1] .+= dt .* (-adv .+ nu .* lap)
        end

        # Poisson Jacobi (vetorizado com buffer)
        function jacobi_psi!(psi, omega, iters)
            dx2 = dx^2; dy2 = dy^2; beta = 1/(2*(dx2+dy2))
            for _ in 1:iters
                set_psi_bcs!(psi)
                psinew[2:end-1,2:end-1] .= beta .* (
                    (psi[3:end,2:end-1] .+ psi[1:end-2,2:end-1]) .* dy2 .+
                    (psi[2:end-1,3:end] .+ psi[2:end-1,1:end-2]) .* dx2 .+
                    (dx2*dy2) .* omega[2:end-1,2:end-1]
                )
                # manter fronteiras a zero no buffer
                psinew[:,1] .= 0.0; psinew[:,end] .= 0.0; psinew[1,:] .= 0.0; psinew[end,:] .= 0.0
                psi, psinew = psinew, psi   # swap
            end
            return psi
        end

        # ------------- loop temporal -------------
        t0 = time(); steps = 0; dt_acc = 0.0; res = Inf
        # warm-up leve
        set_psi_bcs!(psi); update_uv_from_psi!(u,v,psi); update_wall_vorticity!(omega,psi)
        step_vorticity!(omeganew, omega, u, v, 1e-6); psi = jacobi_psi!(psi, omega, 1)

        for it in 1:maxiter
            set_psi_bcs!(psi)
            update_uv_from_psi!(u, v, psi)
            update_wall_vorticity!(omega, psi)

            umax = max(maximum(abs.(u)), 1e-12)
            vmax = max(maximum(abs.(v)), 1e-12)
            dtc1 = dx/umax; dtc2 = dy/vmax
            dtd1 = dx^2/(4nu); dtd2 = dy^2/(4nu)
            dt = CFL * min(dtc1, dtc2, dtd1, dtd2)

            step_vorticity!(omeganew, omega, u, v, dt)

            # alinhar bordas antes do res (para nao contaminar)
            omeganew[1,:] .= omega[1,:];   omeganew[end,:] .= omega[end,:]
            omeganew[:,1] .= omega[:,1];   omeganew[:,end] .= omega[:,end]
            res = maximum(abs.(omeganew .- omega))

            omega .= omeganew
            psi = jacobi_psi!(psi, omega, JIT)

            steps += 1; dt_acc += dt
            if it % 200 == 0
                @printf("Iter %d, dt=%.2e, res=%.2e\n", it, dt, res)
                flush(stdout)
            end
            if res < tol
                @printf("Convergiu em %d passos\n", it)
                flush(stdout)
                break
            end
        end

        elapsed = time() - t0
        cells = (N-2)*(N-2)
        mlups_vort = cells*steps/elapsed/1e6
        avg_dt = dt_acc/max(steps,1)

        @printf("N=%d Re=%.1f steps=%d time=%.3fs MLUPS(vort)=%.2f avg_dt=%.2e res=%.2e\n",
                N, Re, steps, elapsed, mlups_vort, avg_dt, res)

        # ----------- saidas -----------
        data_dir = joinpath(pwd(), "data")
        isdir(data_dir) || mkpath(data_dir)

        open(joinpath(data_dir, "u_center_julia_N$(N).dat"), "w") do f
            for j in 1:N
                @printf(f, "%f %f\n", (j-1)*dy, u[cld(N,2), j])
            end
        end
        open(joinpath(data_dir, "v_center_julia_N$(N).dat"), "w") do f
            for i in 1:N
                @printf(f, "%f %f\n", (i-1)*dx, v[i, cld(N,2)])
            end
        end

        # CSV de performance continua na raiz
        sumfile = "summary_julia.csv"
        newfile = !isfile(sumfile)
        open(sumfile, "a") do f
            if newfile
                write(f, "timestamp,N,Re,CFL,tol,jacobi_iters,steps,time_s,avg_dt,MLUPS_vort,res\n")
            end
            @printf(f, "%s,%d,%.1f,%.3f,%.1e,%d,%d,%.6f,%.6e,%.3f,%.3e\n",
                    Dates.format(now(), dateformat"yyyy-mm-ddTHH:MM:SS"),
                    N, Re, CFL, tol, JIT, steps, elapsed, avg_dt, mlups_vort, res)
        end
    end # @views
end

main()
